File: 20_valid_parentheses/isValid.go
func isValid(s string) bool {
	var stack []rune
	p := map[rune]rune{
		')': '(',
		']': '[',
		'}': '{',
	}

	for _, i := range s {
		if v, ok := p[i]; ok == true {
			if len(stack) > 0 && stack[len(stack)-1] == v {
				stack = stack[:len(stack)-1]
			} else {
				return false
			}
		} else {
			stack = append(stack, i)
		}
	}

	if len(stack) > 0 {
		return false
	}

	return true
}

File: 20_valid_parentheses/isValid.java
class Solution {
    public boolean isValid(String s) {
        Character[] values = { '(', '[', '{' };
        List<Character> stack = new ArrayList<>();
        
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (Arrays.asList(values).contains(c)) {
                stack.add(c);
            } else {
                if (stack.size() == 0) {
                    return false;
                } else {
                    Character current = stack.remove(stack.size() - 1);

                    if (current == '(' && c != ')') {
                        return false;
                    } else if (current == '[' && c != ']') {
                        return false;
                    }
                    if (current == '{' && c != '}') {
                        return false;
                    }
                }
            }
        }

        if (stack.size() > 0) {
            return false;
        }

        return true;
    }
}